sycl/doc/extensions/proposed/sycl_ext_oneapi_if_device.asciidoc

```diff
@@ -0,0 +1,166 @@
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension adds a way to conditionally select code based on whether the
+caller runs on a device or runs on the host.  This allows an application to
+call some common function from either host or device, and then select one path
+or another for each case.  The `+__SYCL_DEVICE_ONLY__+` macro from the core
+SYCL specification provides this same ability, but the new API in this
+extension provides a solution that works even in a 1-pass implementation of
+SYCL.  In such an implementation, the preprocessor runs only once, so there is
+no way to define the `+__SYCL_DEVICE_ONLY__+` macro differently for the host
+and device compilations.
+
+The goal of this extension is to provide a better alternative to the
+`+__SYCL_DEVICE_ONLY__+` macro, adopt it into a future version of the SYCL
+specification, and deprecate usage of that macro.
+
+The APIs added in this extension are designed in the same style as the APIs in
+link:sycl_ext_oneapi_if_device_has.asciidoc[sycl_ext_oneapi_if_device_has] in
+order to provide a consistent style for conditionally selecting code based on
+where it runs.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_IF_DEVICE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New `if_device` and `if_host` free functions
+
+This extension adds two new free functions which may be called from either
+device or host code.
+
+```
+namespace sycl::ext::oneapi::experimental {
+
+template<typename T>
+/* unspecified */ if_device(T fn);
+
+template<typename T>
+/* unspecified */ if_host(T fn);
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+The parameter `fn` must be a C++ `Callable` object which is invocable with an
+empty parameter list.  When called from device code, normal SYCL restrictions
+apply, so `fn` must not be a function pointer or a pointer to a member
+function.  (The expectation is that most applications will pass a lambda
+expression for this parameter.)
```

Comment:
I'm a bit confused by this paragraph.  It seems at odds with the later statement:

> `fn` (and functions it calls) may legally use C++ features that are forbidden in device code.

Why can't `if_host` use a function pointer or a pointer to a member function as the value of `fn`?  We know that it's only going to get called on the host, and the spec as written would already allow a function pointer or pointer to a member function to be used if it was (trivially) wrapped in a lambda.

Reply:
Yes, I think there is no reason to have this restriction for the `if_host` parameter (or the "otherwise" parameter to `if_device`).  I'll loosen the restriction to allow function pointers here.  I still think that most applications will use a lambda, though.

Reply:
Addressed in 9597c9253eeac776f7a32c0b168263781c8e4fa0